src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,171 @@
+ *
+ * @param w width in bits of connection to outside world
+ */
+class TSIHostWidgetIO(val w: Int)(implicit val p: Parameters) extends Bundle {
+  val serial = new SerialIO(w)
+  val debug: SerialIO = if(p(PeripheryTSIHostKey).bypassMMIO) { new SerialIO(w) } else { null } // unconnected in normal operation
+}
+
+/**
+ * I/O bundle to connect to the mmio interaction class
+ */
+trait TLTSIHostMMIOBundle {
+  implicit val p: Parameters
+  val serial = new SerialIO(p(PeripheryTSIHostKey).mmioRegWidth)
+}
+
+/**
+ * Mixin defining the module used to communicate between the MMIO and the parser
+ */
+trait TLTSIHostMMIOModule extends HasRegMap {
+  implicit val p: Parameters
+
+  val io: TLTSIHostMMIOBundle
+
+  val txQueue = Module(new Queue(UInt(p(PeripheryTSIHostKey).mmioRegWidth.W), p(PeripheryTSIHostKey).txQueueEntries)) // where is the queue being dequeued (to the parser serializer)
+  val rxQueue = Module(new Queue(UInt(p(PeripheryTSIHostKey).mmioRegWidth.W), p(PeripheryTSIHostKey).rxQueueEntries)) // where is the queue being enqueued (from the parser deserializer)
+
+  io.serial.out <> txQueue.io.deq
+  rxQueue.io.enq <> io.serial.in
+
+  // memory mapped registers and connections to the queues
+  regmap(
+    TSIHostWidgetCtrlRegs.txQueueOffset -> Seq(RegField.w(p(PeripheryTSIHostKey).mmioRegWidth, txQueue.io.enq)),
+    TSIHostWidgetCtrlRegs.rxQueueOffset -> Seq(RegField.r(p(PeripheryTSIHostKey).mmioRegWidth, rxQueue.io.deq))
+  )
+}
+
+/**
+ * Top level class that uses a register router to connect MMIO to the core. This instantes
+ * the implementation module and bundle
+ *
+ * @param beatBytes amount of bytes to send per beat
+ */
+class TLTSIHostMMIO(val beatBytesIn: Int)(implicit p: Parameters)
+  extends TLRegisterRouter(
+    base = p(PeripheryTSIHostKey).baseAddress,
+    devname = "fesvr-mmio",
+    devcompat = Seq("ucbbar,fesvr-widget"),
+    beatBytes = beatBytesIn)(
+      new TLRegBundle(p(PeripheryTSIHostKey), _)    with TLTSIHostMMIOBundle)(
+      new TLRegModule(p(PeripheryTSIHostKey), _, _) with TLTSIHostMMIOModule)
+
+/**
+ * TSIHostWidget to connect the Front End SerVeR to a target TL module.
+ * Resides on the Rocket-Chip periphery.
+ *
+ * Tx is MMIO -> Queue -> Serializer -> TL
+ * Rx is TL -> Deserializer -> Queue -> MMIO
+ *
+ * @param fesvrParams parameter object
+ */
+class TLTSIHostWidget(val beatBytes: Int)(implicit p: Parameters)
+  extends LazyModule
+{
+  // this should communicate over MMIO to the core (tx and rx) and
+  // should covert the communication from TL (aka TL -> Parse -> MMIO, MMIO -> Parse -> TL)
+  val mmio: TLTSIHostMMIO = if (p(PeripheryTSIHostKey).bypassMMIO) { null } else { LazyModule(new TLTSIHostMMIO(beatBytes)) }
+  // reuse module from Howie to take in a decoupled io tsi and convert to a TL stream
+  val parseToTL = LazyModule(new SerialAdapter)
+  // This converts the TL signals given by the parser into a decoupled stream
+  val serdes = LazyModule(new TLSerdesser(
```

Comment:
There are some assertions here that are likely to cause some issues: https://github.com/ucb-bar/testchipip/blob/4775caf9f23826fed9e2400c48ca0cfb88f9eb8f/src/main/scala/Serdes.scala#L554-L558
This is designed to prevent you from messing up the id bitwidths on-chip, but the requirements may be reversed on the host side.

Reply:
Would it be acceptable to have a boolean to swap the requirements? Or is there a preferred method to do this?

Reply:
Offline: Disable when instantiated on host.